File: receptor.py
#!/usr/bin/env python3
"""
Receptor con Arquitectura de Capas - Laboratorio 2 Segunda Parte
Universidad del Valle de Guatemala - CC3067 Redes
Implementa la arquitectura de capas para el receptor
"""

import sys
import os
import subprocess

# Importar los algoritmos existentes
sys.path.append(os.path.dirname(os.path.abspath(__file__)))
from hammingReceptor import hamming_receiver

class CapaAplicacion:
    """Capa de Aplicación: Interacción con el usuario y manejo de mensajes"""
    
    def solicitar_mensaje(self):
        """Solicita un mensaje codificado para recibir"""
        mensaje = input("Ingrese el mensaje codificado recibido: ")
        print(f"[APLICACIÓN] Mensaje codificado recibido: '{mensaje}'")
        return mensaje
        
    def mostrar_mensaje(self, mensaje):
        """Muestra el mensaje decodificado al usuario"""
        print(f"[APLICACIÓN] Mensaje final: '{mensaje}'")
        return mensaje

class CapaPresentacion:
    """Capa de Presentación: Codificación y decodificación de mensajes"""
    
    def codificar_mensaje(self, mensaje):
        """Codifica un mensaje de texto a binario"""
        # Convierte cada carácter a su representación binaria de 8 bits
        mensaje_binario = ''.join(format(ord(c), '08b') for c in mensaje)
        print(f"[PRESENTACIÓN] Mensaje codificado: {mensaje_binario}")
        return mensaje_binario
        
    def decodificar_mensaje(self, mensaje_binario):
        """Decodifica un mensaje binario a texto"""
        # Agrupa los bits en bytes (8 bits) y convierte cada grupo a un carácter
        mensaje = ""
        for i in range(0, len(mensaje_binario), 8):
            byte = mensaje_binario[i:i+8]
            if len(byte) == 8:  # Asegurarse de que tenemos 8 bits
                mensaje += chr(int(byte, 2))
        print(f"[PRESENTACIÓN] Mensaje decodificado: '{mensaje}'")
        return mensaje

class CapaEnlace:
    """Capa de Enlace: Manejo de integridad y corrección de errores"""
    
    def __init__(self, metodo_deteccion='crc32', metodo_correccion='hamming'):
        """
        Inicializa la capa de enlace con métodos de detección y corrección
        
        Args:
            metodo_deteccion: 'crc32' o 'fletcher'
            metodo_correccion: 'hamming' o 'none'
        """
        self.metodo_deteccion = metodo_deteccion
        self.metodo_correccion = metodo_correccion
        print(f"[ENLACE] Inicializado con detección: {metodo_deteccion}, corrección: {metodo_correccion}")
    
    def verificar_integridad(self, mensaje_recibido):
        """Verifica la integridad del mensaje recibido"""
        print(f"[ENLACE] Verificando integridad con método {self.metodo_deteccion}")
        
        if self.metodo_deteccion == 'crc32':
            try:
                # Llamar al receptor de CRC-32 usando Node.js
                process = subprocess.run(['node', 'crc32Receptor.js', mensaje_recibido], 
                                       capture_output=True, text=True)
                resultado = process.stdout.strip()
                
                if "No se detectaron errores" in resultado:
                    print("[ENLACE] ✅ CRC-32: Integridad verificada")
                    # Extraer el mensaje original (sin el CRC)
                    datos = mensaje_recibido[:-32]  # Quitar los 32 bits del CRC
                    return True, datos
                else:
                    print("[ENLACE] ❌ CRC-32: Error de integridad detectado")
                    return False, None
                
            except Exception as e:
                print(f"[ENLACE] Error al verificar con CRC-32: {e}")
                return False, None
        
        # Si no se especificó un método válido, asumir que no hay errores
        print("[ENLACE] Método no implementado, asumiendo mensaje íntegro")
        return True, mensaje_recibido
    
    def corregir_mensaje(self, mensaje_con_errores):
        """Corrige errores en el mensaje si es posible"""
        print(f"[ENLACE] Intentando corregir errores con método {self.metodo_correccion}")
        
        if self.metodo_correccion == 'hamming':
            try:
                # Usar la implementación del receptor Hamming
                resultado = hamming_receiver(mensaje_con_errores)
                
                if resultado["status"] == "success":
                    print("[ENLACE] ✅ Hamming: No se detectaron errores")
                    return True, resultado["message"]
                    
                elif resultado["status"] == "corrected":
                    print(f"[ENLACE] 🔧 Hamming: Error corregido en posición {resultado['error_position']}")
                    return True, resultado["message"]
                    
                else:
                    print("[ENLACE] ❌ Hamming: Error no corregible detectado")
                    return False, None
                    
            except Exception as e:
                print(f"[ENLACE] Error al corregir con Hamming: {e}")
                return False, None
                
        # Si no se especificó un método válido, no podemos corregir
        print("[ENLACE] Método de corrección no implementado")
        return False, None

def main():
    """Función principal para probar el receptor con arquitectura de capas"""
    print("=== RECEPTOR CON ARQUITECTURA DE CAPAS ===")
    
    # Inicializar las capas
    aplicacion = CapaAplicacion()
    presentacion = CapaPresentacion()
    enlace = CapaEnlace(metodo_deteccion='crc32', metodo_correccion='hamming')
    
    while True:
        try:
            # Recibir mensaje codificado (simulado con entrada del usuario)
            mensaje_codificado = aplicacion.solicitar_mensaje()
            
            if mensaje_codificado.lower() == 'quit':
                break
            
            # Verificar integridad del mensaje
            integridad_ok, mensaje_verificado = enlace.verificar_integridad(mensaje_codificado)
            
            if not integridad_ok:
                print("[RECEPTOR] ❌ Mensaje descartado por error de integridad")
                print("="*50)
                continue
                
            # Si hay errores pero son corregibles, intentar corregir
            if self.metodo_correccion != 'none':
                correccion_ok, mensaje_corregido = enlace.corregir_mensaje(mensaje_verificado)
                if correccion_ok:
                    mensaje_verificado = mensaje_corregido
            
            # Decodificar el mensaje (convertir de binario a texto)
            mensaje_decodificado = presentacion.decodificar_mensaje(mensaje_verificado)
            
            # Mostrar el mensaje final
            aplicacion.mostrar_mensaje(mensaje_decodificado)
            
            print("\n[RECEPTOR] Procesamiento completado:")
            print(f"   - Recibido: {mensaje_codificado}")
            print(f"   - Verificado: {mensaje_verificado}")
            print(f"   - Decodificado: {mensaje_decodificado}")
            print("="*50)
            
        except KeyboardInterrupt:
            print("\nSaliendo del receptor...")
            break
        except Exception as e:
            print(f"Error: {e}")

if __name__ == "__main__":
    main()

File: receptor.js
#!/usr/bin/env node

/**
 * Receptor - Arquitectura de Capas con Detección y Corrección de Errores
 * Universidad del Valle de Guatemala - CC3067 Redes
 */
const net = require('net');
const fs = require('fs');
const { execSync } = require('child_process');
const { crc32Receiver } = require('./crc32Receptor.js');

// Puerto en el que escuchar las conexiones
const PUERTO = 8888;

/**
 * Capa de Aplicación
 * Servicios: mostrar_mensaje
 */
class AplicacionLayer {
    /**
     * Muestra el mensaje recibido al usuario
     * @param {string} mensaje - Mensaje decodificado
     * @param {boolean} hayError - Indica si hubo errores no corregibles
     */
    mostrar_mensaje(mensaje, hayError) {
        console.log("\n----- CAPA DE APLICACIÓN: MOSTRAR MENSAJE -----");
        
        if (hayError) {
            console.log("ERROR: Se detectaron errores que no pudieron ser corregidos.");
            console.log("El mensaje puede estar corrupto.");
        } else {
            console.log("Mensaje recibido correctamente:");
            console.log(mensaje);
        }
    }
}

/**
 * Capa de Presentación
 * Servicios: decodificar_mensaje
 */
class PresentacionLayer {
    /**
     * Decodifica un mensaje de ASCII binario a texto
     * @param {string} binario - Mensaje en formato ASCII binario
     * @param {boolean} hayError - Indica si hubo errores no corregibles
     * @returns {string} Mensaje decodificado o cadena vacía si hay errores
     */
    decodificar_mensaje(binario, hayError) {
        console.log("\n----- CAPA DE PRESENTACIÓN: DECODIFICAR MENSAJE -----");
        
        // Si hay errores y no se pueden corregir, no decodificamos
        if (hayError) {
            console.log("No se puede decodificar debido a errores no corregibles.");
            return "";
        }
        
        let texto = "";
        
        // Procesar cada byte (8 bits)
        for (let i = 0; i < binario.length; i += 8) {
            const byte = binario.slice(i, i + 8);
            if (byte.length === 8) {
                const caracter = String.fromCharCode(parseInt(byte, 2));
                texto += caracter;
                console.log(`Binario: ${byte} -> ASCII: ${parseInt(byte, 2)} -> Carácter: '${caracter}'`);
            }
        }
        
        console.log(`Mensaje decodificado: ${texto}`);
        return texto;
    }
}

/**
 * Capa de Enlace
 * Servicios: verificar_integridad, corregir_mensaje
 */
class EnlaceLayer {
    /**
     * Verifica la integridad del mensaje recibido y corrige errores si es posible
     * @param {string} mensaje - Mensaje binario con información de integridad
     * @param {string} algoritmo - Algoritmo utilizado (crc32 o hamming)
     * @returns {Object} Objeto con el mensaje original y si hubo errores no corregibles
     */
    verificar_integridad(mensaje, algoritmo) {
        console.log(`\n----- CAPA DE ENLACE: VERIFICAR INTEGRIDAD (${algoritmo}) -----`);
        
        // Información de la verificación
        let resultado = {
            mensaje_original: "",    // Mensaje sin la información de integridad
            hubo_errores: false,     // ¿Se detectaron errores?
            errores_corregidos: false, // ¿Se pudieron corregir los errores?
            hay_error_no_corregible: false  // ¿Quedaron errores sin corregir?
        };
        
        if (algoritmo === "crc32") {
            // Algoritmo CRC-32 para detección de errores
            console.log("Verificando con CRC-32...");
            
            // Los últimos 32 bits son el CRC
            const datos = mensaje.slice(0, -32);
            const crc_recibido = mensaje.slice(-32);
            
            console.log(`Datos: ${datos}`);
            console.log(`CRC recibido: ${crc_recibido}`);
            
            // Verificar la integridad usando la función de crc32Receptor.js
            const verificacion = crc32Receiver(mensaje);
            
            resultado.mensaje_original = datos;
            resultado.hubo_errores = !verificacion.integridadOk;
            resultado.hay_error_no_corregible = resultado.hubo_errores; // CRC-32 no corrige
            
            console.log(`Integridad: ${verificacion.integridadOk ? "OK" : "ERROR"}`);
            
            if (resultado.hubo_errores) {
                console.log("Se detectaron errores en los datos (CRC no coincide).");
                console.log("CRC-32 no puede corregir errores, solo detectarlos.");
            } else {
                console.log("No se detectaron errores en los datos.");
            }
            
        } else if (algoritmo === "hamming") {
            // Algoritmo de Hamming para corrección de errores
            console.log("Verificando con código Hamming...");
            
            // Guardar mensaje en un archivo temporal para llamar al script Python
            fs.writeFileSync("temp_mensaje_receptor.txt", mensaje);
            
            try {
                // Llamar a hammingReceptor.py usando Python
                const salida = execSync(
                    'python3 hammingReceptor.py --mensaje "$(cat temp_mensaje_receptor.txt)" --jsonOutput',
                    { encoding: 'utf8' }
                );
                
                // Parsear el resultado JSON
                const resultadoHamming = JSON.parse(salida);
                
                resultado.mensaje_original = resultadoHamming.mensajeOriginal;
                resultado.hubo_errores = resultadoHamming.errorDetectado;
                resultado.errores_corregidos = resultadoHamming.errorCorregido;
                resultado.hay_error_no_corregible = resultadoHamming.errorDetectado && !resultadoHamming.errorCorregido;
                
                if (resultado.hubo_errores) {
                    if (resultado.errores_corregidos) {
                        console.log(`Se detectó un error en la posición ${resultadoHamming.posicionError} y fue corregido.`);
                    } else {
                        console.log("Se detectaron múltiples errores que no pudieron ser corregidos.");
                    }
                } else {
                    console.log("No se detectaron errores en los datos.");
                }
                
            } catch (error) {
                console.error("Error al ejecutar hammingReceptor.py:", error.message);
                resultado.hay_error_no_corregible = true;
            }
        }
        
        return resultado;
    }
    
    /**
     * Intenta corregir errores en el mensaje si el algoritmo lo permite
     * @param {Object} resultado - Resultado de la verificación
     * @returns {string} Mensaje corregido o original según corresponda
     */
    corregir_mensaje(resultado) {
        console.log("\n----- CAPA DE ENLACE: CORREGIR MENSAJE -----");
        
        if (!resultado.hubo_errores) {
            console.log("No se detectaron errores, no es necesario corregir.");
            return resultado.mensaje_original;
        }
        
        if (resultado.errores_corregidos) {
            console.log("Errores corregidos exitosamente.");
            return resultado.mensaje_original;
        }
        
        console.log("No fue posible corregir todos los errores.");
        return resultado.mensaje_original;
    }
}

/**
 * Capa de Transmisión
 * Servicios: recibir_informacion
 */
class TransmisionLayer {
    constructor() {
        this.server = null;
    }
    
    /**
     * Inicializa el servidor para recibir información
     * @param {Function} callback - Función a llamar cuando se reciba información
     */
    recibir_informacion(callback) {
        console.log("\n----- CAPA DE TRANSMISIÓN: RECIBIR INFORMACIÓN -----");
        console.log(`Iniciando servidor en el puerto ${PUERTO}...`);
        
        this.server = net.createServer((socket) => {
            console.log(`Cliente conectado: ${socket.remoteAddress}:${socket.remotePort}`);
            
            // Buffer para ir acumulando los datos
            let buffer = "";
            
            // Evento para recibir datos
            socket.on('data', (data) => {
                const chunk = data.toString();
                buffer += chunk;
                console.log(`Recibidos ${chunk.length} bits`);
            });
            
            // Cuando se cierra la conexión, procesamos el mensaje completo
            socket.on('end', () => {
                console.log(`Conexión cerrada. Mensaje completo recibido (${buffer.length} bits).`);
                
                // Solicitar al cliente que indique qué algoritmo se utilizó
                const algoritmo = process.argv[2] || promptAlgoritmo();
                
                // Enviar mensaje de confirmación antes de procesar
                socket.write("Mensaje recibido", () => {
                    // Llamar al callback con el mensaje y el algoritmo
                    callback(buffer, algoritmo);
                });
            });
            
            // Manejar errores de conexión
            socket.on('error', (err) => {
                console.error(`Error de conexión: ${err.message}`);
            });
        });
        
        // Iniciar el servidor
        this.server.listen(PUERTO, () => {
            console.log(`Servidor escuchando en el puerto ${PUERTO}`);
        });
        
        // Manejar errores del servidor
        this.server.on('error', (err) => {
            console.error(`Error del servidor: ${err.message}`);
            if (err.code === 'EADDRINUSE') {
                console.error(`El puerto ${PUERTO} está en uso. Intente con otro puerto.`);
            }
        });
    }
    
    /**
     * Cierra el servidor
     */
    cerrar() {
        if (this.server) {
            this.server.close();
            console.log("Servidor cerrado.");
        }
    }
}

/**
 * Solicita al usuario que indique el algoritmo utilizado
 * @returns {string} Nombre del algoritmo ("crc32" o "hamming")
 */
function promptAlgoritmo() {
    console.log("\nIndique el algoritmo utilizado por el emisor:");
    console.log("1. CRC-32");
    console.log("2. Hamming");
    
    const readline = require('readline-sync');
    const opcion = readline.question("Seleccione (1 o 2): ");
    
    if (opcion === "1") {
        return "crc32";
    } else if (opcion === "2") {
        return "hamming";
    } else {
        console.log("Opción inválida. Usando CRC-32 por defecto.");
        return "crc32";
    }
}

/**
 * Función principal
 */
function main() {
    console.log("\n============= RECEPTOR - ARQUITECTURA DE CAPAS =============");
    
    // Verificar si se especificó el algoritmo como argumento
    let algoritmoSeleccionado = null;
    if (process.argv.length > 2) {
        const arg = process.argv[2].toLowerCase();
        if (arg === "crc32" || arg === "hamming") {
            algoritmoSeleccionado = arg;
            console.log(`Algoritmo seleccionado por línea de comandos: ${algoritmoSeleccionado}`);
        }
    }
    
    // Si no se especificó por línea de comandos, preguntar al usuario
    if (!algoritmoSeleccionado) {
        algoritmoSeleccionado = promptAlgoritmo();
    }
    
    // Inicializar las capas
    const transmision = new TransmisionLayer();
    const enlace = new EnlaceLayer();
    const presentacion = new PresentacionLayer();
    const aplicacion = new AplicacionLayer();
    
    console.log(`\nReceptor configurado para usar el algoritmo: ${algoritmoSeleccionado}`);
    console.log("Esperando mensajes...");
    
    // Iniciar la recepción de mensajes
    transmision.recibir_informacion((mensaje, algoritmo) => {
        console.log(`\nMensaje recibido (${mensaje.length} bits) usando ${algoritmoSeleccionado}`);
        
        try {
            // Verificar integridad (Enlace)
            const resultadoVerificacion = enlace.verificar_integridad(mensaje, algoritmoSeleccionado);
            
            // Corregir errores si es posible (Enlace)
            const mensajeProcesado = enlace.corregir_mensaje(resultadoVerificacion);
            
            // Decodificar mensaje (Presentación)
            const mensajeDecodificado = presentacion.decodificar_mensaje(
                mensajeProcesado, 
                resultadoVerificacion.hay_error_no_corregible
            );
            
            // Mostrar mensaje al usuario (Aplicación)
            aplicacion.mostrar_mensaje(
                mensajeDecodificado, 
                resultadoVerificacion.hay_error_no_corregible
            );
            
        } catch (error) {
            console.error("Error al procesar el mensaje:", error);
        }
        
        console.log("\n" + "=".repeat(60));
        console.log("Esperando nuevo mensaje...");
    });
    
    // Manejar señal de interrupción (Ctrl+C)
    process.on('SIGINT', () => {
        console.log("\nCerrando receptor...");
        transmision.cerrar();
        process.exit(0);
    });
}

// Iniciar el programa
main();
